content/blog/open-source-monitoring-at-the-hpe-customer-technology-center-böblingen.md

```diff
@@ -0,0 +1,377 @@
+
+## Main routine
+
+As a result, the Python main routine of the ILO Prometheus exporter is as simple as (I removed some of the log commands to improve the readability):
+
+```python
+if __name__ == "__main__":
+    input = getServerList()    
+    # Get the monitoring URLs of the server
+    monitor_urls = get_server_urls(input['user'], input['password'], input['server'], input['lfile'])
+    # Start the http_server and the counters, gauges
+    start_http_server(input['port'])
+    c = Counter('ilorest_sample','ILO REST sample number')
+    node = Gauge('ilorest_node','ILO Node Data',['nodename','rack','nodemetric','metricdetail'])
+    delta = Gauge('ConnectorRuntime','Time required for last data collection in seconds')
+    inode = Info('node','Additional Node Info',['node'])
+    # Start the endless loop
+    while True: 
+        t0 = time.time()
+        start0 = t0
+        c.inc()      
+        for server in monitor_urls:
+            try:
+                server_metrics = get_server_data(input['user'], input['password'], monitor_urls[server], input['lfile'])
+                display_results(node, inode, server_metrics, monitor_urls[server])
+            except Exception as ex:
+                log=logopen(input['lfile'])
+                logwriter(log,'Exception')
+                logwriter(log,str(ex.__context__))
+                logclose(log)
+                pass
+        t1 = time.time()
+        delta.set((t1-t0))
+        while ((t1-t0) < input['mintervall']):
+            time.sleep(1.0)
+            t1 = time.time()   
+        # update once per day the input
+        if ( t1 - start0 > 86400):
+            start0 = t1
+            input = getServerList() 
+            monitor_urls = get_server_urls(input['user'], input['password'], input['server'], input['lfile'])
+```
+
+As you can see, I used four Python functions within the main routine:
+
+* *getServerList*
+* *get\_server\_urls*
+* *get\_server\_data*
+* *display\_results*
+
+The first one - *getServerList* - is just reading the input data out of an XML file. Since this is a pretty standard task in Python I do not want to discuss this function here further. It should be sufficient to know, that it is reading the list of server IPs and additional runtime parameters (username, password, monitoring intervall, …) into a Python directory, that is used in the subsequent parts of the script. I want to focus on the other three functions, that are more specific to the ILO Prometheus exporter.
+
+## g﻿et\_server\_urls
+
+The get\_server_urls routine is using the list of ILO IP addresses to gather the correct URLs out of the Redfish resource directory. If the ILO is alive  a Redfish client object for this ILO is created and the redfish resource directory is read. The client object is stored together with the URLs for the thermal, power and computer system information in the python directory server\_urls, which is the return value of the routine.
```

Comment:
I was wondering why some of this paragraph is in italics and other pieces are not. Also, should "Redfish" have an initial capital letter in each case; i.e. "the Redfish resource directory" as well?
![image](https://github.com/hpe-dev-incubator/hpe-dev-portal/assets/77496261/6fe73755-cb75-4d46-aec6-e20ebfd10022)

One other thing... although it's okay grammatically at the beginning, it would be easier to read and more correct to say "The get\server_urls routing **uses** the list..." I noticed that you use this construct in the next few paragraphs as well. So, instead of saying "is using" or "is taking" just say "uses" or "takes"